src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancies/AssetOccupancyManager.cs

```diff
@@ -0,0 +1,385 @@
+            }
+
+            // get bookable period
+            var bookablePeriods = effectiveSchedule.CalculateBookablePeriods(searchDate, periods);
+
+            // remove periods which cannot be occupied
+            bookablePeriods.RemoveAll(x =>
+                !effectiveSchedule.TimeInAdvance.CanOccupy(searchDate + x.StartingTime, bookingDateTime));
+
+            result.AddRange(bookablePeriods);
+        }
+
+        return Task.FromResult(result);
+    }
+
+    protected virtual async Task<List<Schedule>> CalculateEffectiveSchedulesAsync(
+        DateTime startingDateTime,
+        DateTime endingDateTime,
+        List<AssetSchedule> appliedAssetSchedules,
+        Asset asset,
+        AssetCategory category)
+    {
+        var effectiveSchedules = new List<Schedule>();
+
+        foreach (var assetSchedule in appliedAssetSchedules
+                     .Where(x => x.PeriodUsable == PeriodUsable.Accept))
+        {
+            effectiveSchedules.Add(new Schedule
+            {
+                PeriodUsable = PeriodUsable.Accept,
+                StartingDateTime = assetSchedule.StartingDateTime <= startingDateTime
+                    ? startingDateTime
+                    : assetSchedule.StartingDateTime,
+                EndingDateTime = assetSchedule.EndingDateTime >= endingDateTime
+                    ? endingDateTime
+                    : assetSchedule.EndingDateTime,
+                TimeInAdvance = assetSchedule.TimeInAdvance
+            });
+        }
+
+        foreach (var assetSchedule in appliedAssetSchedules
+                     .Where(x => x.PeriodUsable == PeriodUsable.Reject))
+        {
+            // override accept policy
+            var rejectSchedule = new Schedule
+            {
+                PeriodUsable = PeriodUsable.Reject,
+                StartingDateTime = assetSchedule.StartingDateTime <= startingDateTime
+                    ? startingDateTime
+                    : assetSchedule.StartingDateTime,
+                EndingDateTime = assetSchedule.EndingDateTime >= endingDateTime
+                    ? endingDateTime
+                    : assetSchedule.EndingDateTime,
+                TimeInAdvance = assetSchedule.TimeInAdvance
+            };
+
+            var conflictedSchedule = effectiveSchedules.FirstOrDefault(x =>
+                x.IsTimeRangeIntersected(rejectSchedule.StartingDateTime, rejectSchedule.EndingDateTime));
+            if (conflictedSchedule is not null)
+            {
+                effectiveSchedules.Remove(conflictedSchedule);
+                // split conflicted schedule
+                effectiveSchedules.AddRange(
+                    conflictedSchedule.RemoveIntersectingTimeRange(
+                        rejectSchedule.StartingDateTime,
+                        rejectSchedule.EndingDateTime));
+            }
+
+            effectiveSchedules.Add(rejectSchedule);
+        }
+
+        effectiveSchedules.RemoveAll(x => x.IsEmptyTimeRange());
```

Comment:
I think `10:00-10:00` is a valid schedule.

Reply:
I think `10:00(Include)~10:00(Exclude)` is not a valid schedule, and probably `10:00:00~10:00:00:001` is a valid schedule to represent exact time of 10:00:00.

Reply:
Think about `[10:00, 10:00)`. We have used this rule in other entities.

Reply:
> Think about `[10:00, 10:00)`. We have used this rule in other entities.

https://en.wikipedia.org/wiki/Interval_(mathematics)#Classification_of_intervals

`[10:00, 10:00)` means empty interval. So you are right that the start point should not be equal to the end point.